### PR TITLE
RAC-4329 - UCS-Pollers

### PR DESCRIPTION
**Pull request by bbcyyb**
## Background 
We need to update catalog data that is updating for UCS nodes. Pollers are required to achieve this.
#### Items we should include (but not limited to):
- Power/thermal(Temperature)
- Fan status
- PSU status
- Disk status
- LED status
- SEL( if possible)
#### AC:
- Poller job is created
- Feature is validated.
## Detail
- Add new poller type named **Poller.Ucs**
- Create new task and append to Ucs Discovery Graph
- Create new Job which implement business logic
  - Request Ucs relevant data from Ucs-Service
  - Parse and generate standard poller data output
- Write Unit Test

## UnitTest
Passed.

Paired with @pengz1 

@iceiilin @anhou @mcgG

### CODE REVIEW

**Comment by bbcyyb** [comment]
This PR is depended by: https://github.com/RackHD/on-taskgraph/pull/286, https://github.com/RackHD/on-tasks/pull/511
Jenkins depends on: https://github.com/RackHD/ucs-service/pull/24

File: lib/jobs/base-job.js [hunk 1]
```diff
@@ -666,6 +662,67 @@ function baseJobFactory(
...
+    };
+
+    /**
+     * @function _publishUcsCommandResult
+     * @description publish message which has been processed by ucs job to amqp.
+     **/
+    BaseJob.prototype._publishUcsCommandResult = function(uuid, command, result) {
+        return messenger.publish(Constants.Protocol.Exchanges.Task.Name,
+        ['ucs', 'command', command, 'result', uuid].join('.'),
+        new Result({value: result})
```
Comment (pengz1) [inline lib/jobs/base-job.js]: Indent problem for above 3 lines.

File: lib/jobs/base-job.js [hunk 2]
```diff
@@ -666,6 +662,67 @@ function baseJobFactory(
             self.subscriptions.push(subscription);
         });
     };
-    
+
+    /**
+     * @function _subscribeUcsCommand
+     * @description subscribe and wait for a request from the ucs command exchange.
+     **/
+    BaseJob.prototype._subscribeUcsCommand = function(uuid, callback) {
```
Comment (iceiilin) [inline lib/jobs/base-job.js]: If paired with _publishRunUcsCommand, suggest to change this function name to _subscribe**Run**UcsCommand

File: lib/jobs/create-default-pollers.js [hunk 3]
```diff
@@ -83,7 +83,7 @@ function createDefaultPollersJobFactory(
      */
     CreateDefaultPollersJob.prototype.createUcsPoller = function(nodeIds, poller){
         var self = this;
-        return Promise.map(nodeIds, function(nodeId) {
+        return Promise.each(nodeIds, function(nodeId) {
```
Comment (iceiilin) [inline lib/jobs/create-default-pollers.js]: Is there any reason to change it from map to each?

File: lib/jobs/ucs-job.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        return waterline.workitems.update(
+            {name: "Pollers.UCS"}, 
+            {failureCount: 0}
+        )
+        .then(function() {
+            return self._subscribeUcsCommand(
+                self.routingKey,
+                self._subscribeUcsCallback.bind(self)
+            );
+        });
```
Comment (iceiilin) [inline lib/jobs/ucs-job.js]: do you need to catch error and exit here?

File: lib/jobs/ucs-job.js [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            return waterline.obms.findByNode(data.node, 'ucs-obm-service', true)
+            .then(function(obmSetting) {
+                if (obmSetting) {
+                    return {
+                        obmSetting: obmSetting,
+                        command: data.config.command,
+                        data: data
+                    };
+                } else {
+                    throw new Error("No ucs-obm-service found for id: %s".format(data.node));
```
Comment (iceiilin) [inline lib/jobs/ucs-job.js]: I would prefer to use promise.reject to make it consist in a promise chain.

File: lib/jobs/create-default-pollers.js (hunk 3 above)
Comment (pengz1) [reply lib/jobs/create-default-pollers.js]: Please refer to: https://rackhd.atlassian.net/browse/RAC-6140

**Comment by panpan0000** [comment]
pretty PR description . +0

File: lib/jobs/base-job.js (hunk 1 above)
Comment (bbcyyb) [reply lib/jobs/base-job.js]: fixed.

File: lib/jobs/base-job.js (hunk 2 above)
Comment (bbcyyb) [reply lib/jobs/base-job.js]: Fixed, use universal naming for subscribe and publich method on base-job.js.

File: lib/jobs/ucs-job.js (hunk 4 above)
Comment (bbcyyb) [reply lib/jobs/ucs-job.js]: Add a new try-catch block after statndard _run logic.

File: lib/jobs/ucs-job.js (hunk 5 above)
Comment (bbcyyb) [reply lib/jobs/ucs-job.js]: Fixed.

**Pull request merged**